feat(getTogglerProps): add 'tabIndex', 'onKeyUp'

tansongyang opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

If you're new to contributing to open source projects, you might find this free
video course helpful: http://kcd.im/pull-request

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->
**What**: Added 'tabIndex' and 'onKeyUp' to `getTogglerProps`.

<!-- Why are these changes necessary? -->
**Why**: See #5

<!-- How were these changes implemented? -->
**How**: 'tabIndex' was straightforward. 'onKeyUp' was not trivial, and I'd like to discuss it further below.

<!-- Have you done all of these things?  -->
**Checklist**:
<!-- add "N/A" to the end of each line that's irrelevant to your changes -->
<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->
- [ ] Documentation
- [x] Tests
- [ ] Ready to be merged <!-- In your opinion, is this ready to be merged as soon as it's reviewed? -->
- [x] Added myself to contributors table <!-- this is optional, see the contributing guidelines for instructions -->

<!-- feel free to add additional comments -->

I think the implementation of `handleKeyUp` needs to be addressed. If we agree on an approach, then I think we should update the documentation.

`handleKeyUp` was supposed to make it easy to build with plain old `<span>`s, and things like that. However, `<button>`s and `<input>`s already handle certain key presses, so I needed to filter those out. The best way I could think of doing this was to check the `tagName` of the target. That's a potentially messy implementation, so I'd like your thoughts on the matter.

If you are OK with this implementation, I wonder if it's worth updating the docs to inform users that they will be getting this functionality for free.

tansongyang committed:
feat(getTogglerProps): add 'tabIndex', 'onKeyUp'

These make it easier to build custom checkboxes, switches, etc.

See #5

kentcdodds commented on src/index.js:
```diff
@@ -29,8 +29,10 @@ class Toggle extends Component {
   getTogglerProps = (props = {}) => ({
     'aria-controls': 'target',
     'aria-expanded': Boolean(this.getOn()),
+    tabIndex: props.tabIndex == null ? 0 : props.tabIndex,
```
Could probably just do: `tabIndex: 0` and the `...props` will override `tabIndex` if it's on there.

tansongyang replied:
Duly noted. I'll update this later today.

kentcdodds commented on src/index.js:
```diff
@@ -57,6 +59,20 @@ class Toggle extends Component {
   toggle = this.setOnState.bind(this, undefined)
 
+  toggleKeys = ['Enter', ' '] // This matches <button> behavior
+  handleKeyUp = event => {
```
I'm ok with this implementation, but why doesn't `onClick` fire in addition to `onKeyUp`? I would expect `this.toggle` to be called twice 🤔

tansongyang replied:
`onClick` does fire as well in certain cases. That's what I wanted input on. I ran into this, so I check the target in the body of `handleKeyUp`. If it's a `<button>`, we do nothing, since buttons respond to both `Enter` and `Space`. If it's an `<input>`, we don't respond to `Space` because browsers will already treat it as a click. However, we will respond to `Enter` to match `<button>`. For any other elements, we respond to both `Enter` and `Space`.

I'm not a _huge_ fan of this, and I worry that there's a missing corner case. However, that's what I came up with. Can this be improved?

kentcdodds replied:
What if we make a new prop getter that has the keyup prop. So then the user can decide. Could be: `getNonButtonTogglerProps` or something. I think the implementation would be simpler, though it'd make the API potentially more confusing... Maybe this is good as is... 🤔

tansongyang replied:
Actually, I think you have a good point. That brings up another idea that I'll propose.

```JavaScript
getInputTogglerKeyUpProp(onKeyUp) {
  return {
    onKeyUp: (event) => {
      onKeyUp(event)
      if (event.key === ' ') {
        this.toggle()
      }
    }
  }
}
```

```JavaScript
getGenericTogglerKeyUpProp(onKeyUp) {
  return {
    onKeyUp: (event) => {
      onKeyUp(event)
      if (toggleKeys.indexOf(event.key) > -1) {
        this.toggle()
      }
    }
  }
}
```

Usage:

```JavaScript
<input type="checkbox" {...getTogglerProps()} {...getInputTogglerKeyUpProp()}>
<span {...getTogglerProps()} {...getGenericTogglerKeyUpProp()}>
```

Basically, `getTogglerProps` is designed for `<button>`s. We add these extra methods to support other elements. We'll just be sure to document.

Thoughts?

kentcdodds replied:
I was thinking that `getInputTogglerKeyUpProp` would actually do:

```javascript
getInputTogglerKeyUpProp(onKeyUp) {
  return this.getTogglerProps({
    onKeyUp: (event) => {
      onKeyUp(event)
      if (event.key === ' ') {
        this.toggle()
      }
    }
  })
}
```

Same with the generic one...

kentcdodds reviewed (commented):
Looking pretty good. I've got a few questions.

tansongyang committed:
Apply code review comments

Introduce `getInputTogglerProps` and `getGenericTogglerProps`.
Set default `tabIndex`.

tansongyang commented:
I've updated based on the suggestions. You can see the `switch` and `checkbox` stories for example usage. I like this API more. If you think this is good, then I believe there should only be two more items.

1. Do you think the names `getInputTogglerProps` and `getGenericTogglerProps` are OK? I'm mildly concerned about `getGenericTogglerProps` because "generic" has a special meaning in types. (I know types aren't in JS, but I thought I'd mention it.) Other things we might consider: `getElementTogglerProps` or `getAllTogglerProps`.
2. Update documentation.

kentcdodds commented on README.md:
```diff
@@ -137,7 +137,7 @@ The function is passed as the child prop:
 |-------------------|-------------|---------------------------|--------------------------------------------------------------------------------------------------|
 | `on`              | state       | `boolean`                 | The current `on` state of toggle                                                                 |
-| `getTogglerProps` | prop getter | `function(props: object)` | returns the props you should apply to the button element you render. Includes `aria-` attributes |
+| `getTogglerProps` | prop getter | `function(props: object)` | returns the props you should apply to the element you render. Includes `aria-` attributes |
```
We'll need to add two items to this table.

tansongyang replied:
Wil do. Also, if it's all the same to you, I'm starting to think `getElementTogglerProps` is a better name than `getGenericTogglerProps`.

kentcdodds replied:
That's fine with me :)

kentcdodds reviewed (commented):
Just some docs updates :smile:

tansongyang committed:
Rename to getElementTogglerProps

tansongyang committed:
Document new props getters

tansongyang commented:
Let me know if there's anything else to be done here!

kentcdodds reviewed (approved):
Thanks a bunch!

kentcdodds commented:
Thanks so much for your help! I've added you as a collaborator on the project. Please make sure that you review the `other/MAINTAINING.md` and `CONTRIBUTING.md` files (specifically the bit about the commit messages and the git hooks) and familiarize yourself with the code of conduct (we're using [the contributor covenant](http://contributor-covenant.org/)). You might also want to [watch the repo](https://help.github.com/articles/watching-repositories/) to be notified when someone files an issue/PR. Please continue to make PRs as you feel the need (you can make your branches directly on the repo rather than your fork if you want). Thanks! And welcome to the team :)

tansongyang commented:
Thank you as well!
